File: clients/python/games/battleship.py
from games.base import BaseGame
from framework.i18n import t

SHIP_SIZES = [5, 4, 3, 3, 2]
GRID = 10


class Battleship(BaseGame):
    min_players = max_players = 2

    def __init__(self):
        self.players = []
        self._phase = 'place'
        self._ships = {}
        self._placed = {}
        self._shots = {}
        self._turn_idx = 0
        self._place_turn = 0

    def start(self, players):
        self.players = players
        self._turn_idx = 0
        self._place_turn = 0
        for p in players:
            self._ships[p] = []
            self._placed[p] = 0
            self._shots[p] = {}

    def current_turn(self):
        if self._phase == 'place':
            return self.players[self._place_turn % len(self.players)]
        return self.players[self._turn_idx]

    def validate_move(self, player_id, move_data):
        if player_id != self.current_turn():
            return False
        if self._phase == 'place':
            pl = move_data.get('place')
            if not pl:
                return False
            idx = self._placed[player_id]
            if idx >= len(SHIP_SIZES):
                return False
            return self._can_place(player_id, idx, pl['row'], pl['col'], pl['horiz'])
        else:
            sh = move_data.get('shot')
            if not sh:
                return False
            key = (sh['row'], sh['col'])
            return (0 <= sh['row'] < GRID and 0 <= sh['col'] < GRID and
                    key not in self._shots[player_id])

    def _can_place(self, pid, ship_idx, r, c, horiz):
        size = SHIP_SIZES[ship_idx]
        cells = [(r, c + i) if horiz else (r + i, c) for i in range(size)]
        if any(not (0 <= cr < GRID and 0 <= cc < GRID) for cr, cc in cells):
            return False
        occupied = {cell for s in self._ships[pid] for cell in s}
        return not any(cell in occupied for cell in cells)

    def apply_move(self, player_id, move_data):
        if self._phase == 'place':
            pl = move_data['place']
            idx = self._placed[player_id]
            size = SHIP_SIZES[idx]
            cells = set(
                (pl['row'], pl['col'] + i) if pl['horiz'] else (pl['row'] + i, pl['col'])
                for i in range(size)
            )
            self._ships[player_id].append(cells)
            self._placed[player_id] += 1
            self._place_turn += 1
            if all(self._placed[p] == len(SHIP_SIZES) for p in self.players):
                self._phase = 'battle'
                self._turn_idx = 0
        else:
            sh = move_data['shot']
            key = (sh['row'], sh['col'])
            opp = self.players[1 - self.players.index(player_id)]
            hit = any(key in ship for ship in self._ships[opp])
            self._shots[player_id][key] = 'hit' if hit else 'miss'
            self._turn_idx = 1 - self._turn_idx

    def _all_sunk(self, pid, shooter):
        return all(ship <= self._shots[shooter].keys() for ship in self._ships[pid])

    def is_over(self):
        if self._phase != 'battle':
            return False, None
        for i, pid in enumerate(self.players):
            opp = self.players[1 - i]
            if self._all_sunk(pid, opp):
                return True, opp
        return False, None

    def render(self, perspective=None):
        if not perspective or perspective not in self.players:
            perspective = self.players[0] if self.players else None
        opp = next((p for p in self.players if p != perspective), None)
        lines = [t('battleship.title', player=perspective)]
        own_cells = {c for s in self._ships.get(perspective, []) for c in s}
        lines.append(t('battleship.own_board'))
        for r in range(GRID):
            row = '  '
            for c in range(GRID):
                key = (r, c)
                opp_shots = self._shots.get(opp, {})
                if key in opp_shots:
                    row += ('X' if opp_shots[key] == 'hit' else 'o') + ' '
                elif (r, c) in own_cells:
                    row += 'S '
                else:
                    row += '. '
            lines.append(row)
        if opp:
            lines.append(t('battleship.enemy_board', opp=opp))
            for r in range(GRID):
                row = '  '
                for c in range(GRID):
                    s = self._shots.get(perspective, {}).get((r, c))
                    row += ('X' if s == 'hit' else 'o' if s == 'miss' else '.') + ' '
                lines.append(row)
        return '\n'.join(lines)

    def get_state(self, perspective=None):
        opp = next((p for p in self.players if p != perspective), None)
        return {
            'phase': self._phase,
            'turn': self.current_turn(),
            'own_ships': [list(c) for s in self._ships.get(perspective, []) for c in s],
            'my_shots': {f'{r},{c}': v for (r, c), v in self._shots.get(perspective, {}).items()},
            'opp_shots': ({f'{r},{c}': v for (r, c), v in self._shots.get(opp, {}).items()}
                          if opp else {}),
        }

File: clients/node/src/games/battleship.js
'use strict';
const { BaseGame } = require('./base');
const { t } = require('../framework/i18n');

const SHIP_SIZES = [5, 4, 3, 3, 2];
const GRID = 10;

class Battleship extends BaseGame {
  constructor() {
    super();
    this.players = [];
    this._phase = 'place';
    this._ships = {};
    this._placed = {};
    this._shots = {};
    this._turnIdx = 0;
    this._placeTurn = 0;
  }

  start(players) {
    this.players = players;
    this._turnIdx = 0;
    this._placeTurn = 0;
    for (const p of players) {
      this._ships[p] = [];
      this._placed[p] = 0;
      this._shots[p] = new Map();
    }
  }

  currentTurn() {
    if (this._phase === 'place') return this.players[this._placeTurn % this.players.length];
    return this.players[this._turnIdx];
  }

  validateMove(playerId, moveData) {
    if (playerId !== this.currentTurn()) return false;
    if (this._phase === 'place') {
      const pl = moveData.place;
      if (!pl) return false;
      const idx = this._placed[playerId];
      if (idx >= SHIP_SIZES.length) return false;
      return this._canPlace(playerId, idx, pl.row, pl.col, pl.horiz);
    } else {
      const sh = moveData.shot;
      if (!sh) return false;
      const key = `${sh.row},${sh.col}`;
      return sh.row >= 0 && sh.row < GRID && sh.col >= 0 && sh.col < GRID &&
             !this._shots[playerId].has(key);
    }
  }

  _canPlace(pid, shipIdx, r, c, horiz) {
    const size = SHIP_SIZES[shipIdx];
    const cells = [];
    for (let i = 0; i < size; i++) {
      cells.push(horiz ? [r, c + i] : [r + i, c]);
    }
    if (cells.some(([cr, cc]) => cr < 0 || cr >= GRID || cc < 0 || cc >= GRID)) return false;
    const occupied = new Set(this._ships[pid].flatMap(s => [...s]));
    return !cells.some(([cr, cc]) => occupied.has(`${cr},${cc}`));
  }

  applyMove(playerId, moveData) {
    if (this._phase === 'place') {
      const pl = moveData.place;
      const idx = this._placed[playerId];
      const size = SHIP_SIZES[idx];
      const cells = new Set();
      for (let i = 0; i < size; i++) {
        const [r, c] = pl.horiz ? [pl.row, pl.col + i] : [pl.row + i, pl.col];
        cells.add(`${r},${c}`);
      }
      this._ships[playerId].push(cells);
      this._placed[playerId]++;
      this._placeTurn++;
      if (this.players.every(p => this._placed[p] === SHIP_SIZES.length)) {
        this._phase = 'battle';
        this._turnIdx = 0;
      }
    } else {
      const sh = moveData.shot;
      const key = `${sh.row},${sh.col}`;
      const opp = this.players[1 - this.players.indexOf(playerId)];
      const hit = this._ships[opp].some(ship => ship.has(key));
      this._shots[playerId].set(key, hit ? 'hit' : 'miss');
      this._turnIdx = 1 - this._turnIdx;
    }
  }

  _allSunk(pid, shooter) {
    return this._ships[pid].every(ship => [...ship].every(cell => this._shots[shooter].has(cell)));
  }

  isOver() {
    if (this._over) return [true, this._winner];
    if (this._phase !== 'battle') return [false, null];
    for (let i = 0; i < this.players.length; i++) {
      const opp = this.players[1 - i];
      if (this._allSunk(this.players[i], opp)) return [true, opp];
    }
    return [false, null];
  }

  render(perspective) {
    if (!perspective || !this.players.includes(perspective)) {
      perspective = this.players[0] || null;
    }
    const opp = this.players.find(p => p !== perspective) || null;
    const lines = [t('battleship.title', { player: perspective })];
    const ownCells = new Set(this._ships[perspective]?.flatMap(s => [...s]) || []);
    lines.push(t('battleship.own_board'));
    for (let r = 0; r < GRID; r++) {
      let row = '  ';
      for (let c = 0; c < GRID; c++) {
        const key = `${r},${c}`;
        const oppShots = opp ? this._shots[opp] : new Map();
        if (oppShots.has(key)) row += (oppShots.get(key) === 'hit' ? 'X' : 'o') + ' ';
        else if (ownCells.has(key)) row += 'S ';
        else row += '. ';
      }
      lines.push(row);
    }
    if (opp) {
      lines.push(t('battleship.enemy_board', { opp }));
      for (let r = 0; r < GRID; r++) {
        let row = '  ';
        for (let c = 0; c < GRID; c++) {
          const s = this._shots[perspective]?.get(`${r},${c}`);
          row += (s === 'hit' ? 'X' : s === 'miss' ? 'o' : '.') + ' ';
        }
        lines.push(row);
      }
    }
    return lines.join('\n');
  }

  getState(perspective) {
    const opp = this.players.find(p => p !== perspective) || null;
    const shotsObj = m => Object.fromEntries([...m]);
    return {
      phase: this._phase,
      turn: this.currentTurn(),
      ownShips: this._ships[perspective] ? this._ships[perspective].flatMap(s => [...s]) : [],
      myShots: shotsObj(this._shots[perspective] || new Map()),
      oppShots: opp ? shotsObj(this._shots[opp] || new Map()) : {},
    };
  }

  loadState(data, perspective) {
    if (!data) return;
    if (data.phase != null) this._phase = data.phase;
    if (data.turn != null) {
      const idx = this.players.indexOf(data.turn);
      if (idx >= 0) {
        if (this._phase === 'place') this._placeTurn = idx;
        else this._turnIdx = idx;
      }
    }
    if (perspective && this.players.includes(perspective)) {
      const opp = this.players.find(p => p !== perspective);
      if (data.ownShips) {
        this._ships[perspective] = [new Set(data.ownShips)];
      }
      if (data.myShots) {
        this._shots[perspective] = new Map(Object.entries(data.myShots));
      }
      if (opp && data.oppShots) {
        this._shots[opp] = new Map(Object.entries(data.oppShots));
      }
    }
  }

  parseInput(raw) {
    const trimmed = raw.trim();
    if (trimmed.startsWith('{')) {
      try { const obj = JSON.parse(trimmed); if (obj && typeof obj === 'object') return obj; } catch (_) {}
    }
    const parts = trimmed.split(/\s+/);
    if (parts.length === 3) {
      const r = Number(parts[0]), c = Number(parts[1]);
      if (!isNaN(r) && !isNaN(c)) {
        return { place: { row: r, col: c, horiz: parts[2].toLowerCase() !== 'v' } };
      }
    }
    if (parts.length === 2) {
      const r = Number(parts[0]), c = Number(parts[1]);
      if (!isNaN(r) && !isNaN(c)) {
        if (this._phase === 'place') return { place: { row: r, col: c, horiz: true } };
        return { shot: { row: r, col: c } };
      }
    }
    return null;
  }

  getHelp() {
    return [
      'Place ships secretly, then take turns calling coordinates to sink them.',
      'Place ship: <row> <col> <h|v>   e.g. "3 4 h"  (or "3 4 v" for vertical)',
      'Shoot:      <row> <col>          e.g. "3 4"',
    ];
  }
}

module.exports = { Battleship };
